TextFieldEffects/TextFieldEffects/HoshiTextField.swift

```diff
@@ -12,178 +12,226 @@ import UIKit
+   */
+  @IBInspectable dynamic open var borderInactiveColor: UIColor? {
+    didSet {
+      updateBorder()
     }
-    
-    private func placeholderFontFromFont(_ font: UIFont) -> UIFont! {
-        let smallerFont = UIFont(name: font.fontName, size: font.pointSize * placeholderFontScale)
-        return smallerFont
+  }
+
+  /**
+   The color of the border when it has content.
+
+   This property applies a color to the lower edge of the control. The default value for this property is a clear color.
+   */
+  @IBInspectable dynamic open var borderActiveColor: UIColor? {
+    didSet {
+      updateBorder()
     }
-    
-    private func rectForBorder(_ thickness: CGFloat, isFilled: Bool) -> CGRect {
-        if isFilled {
-            return CGRect(origin: CGPoint(x: 0, y: frame.height-thickness), size: CGSize(width: frame.width, height: thickness))
-        } else {
-            return CGRect(origin: CGPoint(x: 0, y: frame.height-thickness), size: CGSize(width: 0, height: thickness))
-        }
+  }
+
+  /**
+   The color of the placeholder text.
+
+   This property applies a color to the complete placeholder string. The default value for this property is a black color.
+   */
+  @IBInspectable dynamic open var placeholderColor: UIColor = .black {
+    didSet {
+      updatePlaceholder()
     }
-    
-    private func layoutPlaceholderInTextRect() {        
-        let textRect = self.textRect(forBounds: bounds)
-        var originX = textRect.origin.x
-        switch self.textAlignment {
-        case .center:
-            originX += textRect.size.width/2 - placeholderLabel.bounds.width/2
-        case .right:
-            originX += textRect.size.width - placeholderLabel.bounds.width
-        default:
-            break
-        }
-        placeholderLabel.frame = CGRect(x: originX, y: textRect.height/2,
-            width: placeholderLabel.bounds.width, height: placeholderLabel.bounds.height)
-        activePlaceholderPoint = CGPoint(x: placeholderLabel.frame.origin.x, y: placeholderLabel.frame.origin.y - placeholderLabel.frame.size.height - placeholderInsets.y)
+  }
+
+  /**
+   The scale of the placeholder font.
+
+   This property determines the size of the placeholder label relative to the font size of the text field.
+   */
+  var placeholderFontSize: CGFloat = 12  {
+    didSet {
+      updatePlaceholder()
+    }
+  }
+  
+  override open var placeholder: String? {
+    didSet {
+      updatePlaceholder()
+    }
+  }
+
+  override open var bounds: CGRect {
+    didSet {
+      frame = CGRect(x: frame.origin.x,
+                     y: frame.origin.y,
+                     width: bounds.size.width,
+                     height: height)
+    }
+  }
+
+  override open func awakeFromNib() {
+    super.awakeFromNib()
+    initPlaceholderFont()
+  }
+
+  private let borderThickness: (active: CGFloat, inactive: CGFloat) = (active: 2, inactive: 0.5)
```

Comment:
I think you can break it into two separate properties like `activeBorderThickness` and `inactiveBorderThickness`, since there are two properties for active and inactive states. What do you think?

Reply:
are you refering to `inactiveBorderLayer` and `activeBorderLayer`?
How do you see breaking this tuple into to properties helping? readability?

Reply:
No, I meant just `borderThickness`.  Since this is the only property using tuple and readability wise it will be good as well.

Reply:
In my opinion is not a bad design to have a touple for two mutually states, and thinking about it, I probably should have one  struct with a whole configuration for boths states... what do you think?

Reply:
Yeah, `struct` is more verbose compared to `tuple` but looking at all other inactive property I thought why not separate this one into the active and inactive property as well. So if it quick change then go ahead with it.